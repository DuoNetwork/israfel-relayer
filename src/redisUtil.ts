// import Redis from 'ioredis';
// import * as CST from '../common/constants';
// import { IOrderBook, ITrade } from '../common/types';
// import util from './util';

// export class RedisUtil {
// 	private redisPub: Redis.Redis | null = null;
// 	private redisSub: Redis.Redis | null = null;
// 	private handleOrderBooksUpdate:
// 		| ((channel: string, orderBooks: { [src: string]: IOrderBook }) => any)
// 		| null = null;
// 	private handleTradesUpdate: ((channel: string, trades: ITrade[]) => any) | null = null;

// 	public init(redisKey: { host: string; password: string; servername: string }) {
// 		this.redisPub = new Redis(6379, redisKey.host, {
// 			password: redisKey.password,
// 			tls: { servername: redisKey.servername }
// 		});
// 		this.redisSub = new Redis(6379, redisKey.host, {
// 			password: redisKey.password,
// 			tls: { servername: redisKey.servername }
// 		});
// 		this.redisSub.on('message', (channel, message) => this.onMessage(channel, message));
// 		this.redisSub.on('pmessage', (pattern, channel, message) =>
// 			this.onMessage(channel, message, pattern)
// 		);
// 	}

// 	public onMessage(channel: string, message: string, pattern: string = '') {
// 		util.logDebug(pattern + channel);
// 		const [type, ...others] = channel.split('|');
// 		switch (type) {
// 			case CST.DB_ORDER_BOOKS:
// 				if (this.handleOrderBooksUpdate)
// 					this.handleOrderBooksUpdate(channel, { [others[0]]: JSON.parse(message) });
// 				break;
// 			case CST.DB_OB_OVERVIEW:
// 			case CST.DB_OB_GLOBAL:
// 				if (this.handleOrderBooksUpdate)
// 					this.handleOrderBooksUpdate(channel, JSON.parse(message));
// 				break;
// 			case CST.DB_TRADES:
// 				if (this.handleTradesUpdate) this.handleTradesUpdate(channel, JSON.parse(message));
// 				break;
// 			default:
// 				break;
// 		}
// 	}

// 	public onOrderBooks(
// 		handleOrderBooksUpdate: (channel: string, orderBooks: { [src: string]: IOrderBook }) => any
// 	) {
// 		this.handleOrderBooksUpdate = handleOrderBooksUpdate;
// 	}

// 	public onTrades(handleTradesUpdate: (channel: string, trades: ITrade[]) => any) {
// 		this.handleTradesUpdate = handleTradesUpdate;
// 	}

// 	public publish(channel: string, msg: string): void {
// 		if (this.redisPub) this.redisPub.publish(channel, msg);
// 	}

// 	public subscribe(channel: string): void {
// 		if (this.redisSub) this.redisSub.subscribe(channel);
// 	}

// 	public patternSubscribe(pattern: string): void {
// 		if (this.redisSub) this.redisSub.psubscribe(pattern);
// 	}

// 	public unsubscribe(channel: string): void {
// 		if (this.redisSub) this.redisSub.unsubscribe(channel);
// 	}

// 	public patternUnsubscribe(pattern: string): void {
// 		if (this.redisSub) this.redisSub.punsubscribe(pattern);
// 	}

// 	public push(key: string, ...values: string[]): void {
// 		if (this.redisSub) this.redisSub.lpush(key, values);
// 	}

// 	public putBack(key: string, ...values: string[]): void {
// 		if (this.redisSub) this.redisSub.rpush(key, values);
// 	}

// 	public async pop(key: string): Promise<string> {
// 		if (this.redisSub) return await this.redisSub.rpop(key);
// 		return '';
// 	}

// 	public async getQueueLength(key: string): Promise<number> {
// 		if (this.redisSub) return this.redisSub.llen(key);
// 		return 0;
// 	}
// }

// const redisUtil = new RedisUtil();
// export default redisUtil;
