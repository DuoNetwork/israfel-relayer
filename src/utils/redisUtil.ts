import Redis from 'ioredis';
import * as CST from '../common/constants';
import {  IOrderBookSnapshotUpdate, IOrderQueueItem } from '../common/types';
import util from './util';

class RedisUtil {
	private redisPub: Redis.Redis | null = null;
	private redisSub: Redis.Redis | null = null;
	private handleOrderBookUpdate:
		| ((channel: string, orderBookUpdate: IOrderBookSnapshotUpdate) => any)
		| null = null;
	private handleOrderUpdate:
		| ((channel: string, orderQueueItem: IOrderQueueItem) => any)
		| null = null;

	public init(redisKey: { host: string; password: string; servername: string }) {
		this.redisPub = new Redis(6379, redisKey.host, {
			password: redisKey.password,
			tls: { servername: redisKey.servername }
		});
		this.redisSub = new Redis(6379, redisKey.host, {
			password: redisKey.password,
			tls: { servername: redisKey.servername }
		});
		this.redisSub.on('message', (channel, message) => this.onMessage(channel, message));
		this.redisSub.on('pmessage', (pattern, channel, message) =>
			this.onMessage(channel, message, pattern)
		);
	}

	public onMessage(channel: string, message: string, pattern: string = '') {
		util.logDebug(pattern + channel + message);
		const type = channel.split('|')[0];
		switch (type) {
			case CST.DB_ORDER_BOOKS:
				if (this.handleOrderBookUpdate)
					this.handleOrderBookUpdate(channel, JSON.parse(message));
				break;
			case CST.DB_ORDERS:
				if (this.handleOrderUpdate) this.handleOrderUpdate(channel, JSON.parse(message));
				break;
			default:
				break;
		}
	}

	public onOrderBookUpdate(
		handleOrderBookUpdate: (channel: string, orderBookUpdate: IOrderBookSnapshotUpdate) => any
	) {
		this.handleOrderBookUpdate = handleOrderBookUpdate;
	}
	public onOrderUpdate(
		handleOrderUpdate: (channel: string, orderQueueItem: IOrderQueueItem) => any
	) {
		this.handleOrderUpdate = handleOrderUpdate;
	}

	public publish(channel: string, msg: string) {
		if (this.redisPub) return this.redisPub.publish(channel, msg);
		return Promise.resolve(0);
	}

	public increment(key: string) {
		if (this.redisPub) return this.redisPub.incr(key);
		return Promise.reject();
	}

	public set(key: string, value: string) {
		if (this.redisPub) return this.redisPub.set(key, value);
		return Promise.resolve('');
	}

	public get(key: string) {
		if (this.redisPub) return this.redisPub.get(key);
		return Promise.resolve('');
	}

	public hashSet(key: string, field: string, value: string) {
		if (this.redisPub) return this.redisPub.hset(key, field, value);
		return Promise.resolve(0);
	}

	public hashGet(key: string, field: string): Promise<string | null> {
		if (this.redisPub) return this.redisPub.hget(key, field);
		return Promise.resolve(null);
	}

	public async hashMultiGet(key: string, ...fields: string[]) {
		if (this.redisPub) {
			const values: Array<string | null> = await this.redisPub.hmget(key, ...fields);
			const output: { [field: string]: string | null } = {};
			fields.forEach((f, i) => (output[f] = values[i]));
			return output;
		}
		return Promise.resolve({});
	}

	public hashGetAll(key: string): Promise<any> {
		if (this.redisPub) return this.redisPub.hgetall(key);
		return Promise.resolve(null);
	}

	public hashDelete(key: string, field: string) {
		if (this.redisPub) return this.redisPub.hdel(key, field);
	}

	public subscribe(channel: string) {
		if (this.redisSub) this.redisSub.subscribe(channel);
	}

	public patternSubscribe(pattern: string) {
		if (this.redisSub) this.redisSub.psubscribe(pattern);
	}

	public unsubscribe(channel: string) {
		if (this.redisSub) this.redisSub.unsubscribe(channel);
	}

	public patternUnsubscribe(pattern: string) {
		if (this.redisSub) this.redisSub.punsubscribe(pattern);
	}

	public push(key: string, ...values: string[]) {
		if (this.redisPub) this.redisPub.lpush(key, values);
	}

	public putBack(key: string, ...values: string[]) {
		if (this.redisPub) this.redisPub.rpush(key, values);
	}

	public pop(key: string) {
		if (this.redisPub) return this.redisPub.rpop(key);
		return Promise.resolve('');
	}

	public multi() {
		if (this.redisPub) return this.redisPub.multi({ pipeline: false });
		return Promise.resolve('');
	}

	public exec() {
		if (this.redisPub) return this.redisPub.exec();
		return Promise.resolve('');
	}

	public getQueueLength(key: string) {
		if (this.redisPub) return this.redisPub.llen(key);
		return Promise.resolve(0);
	}
}

const redisUtil = new RedisUtil();
export default redisUtil;
