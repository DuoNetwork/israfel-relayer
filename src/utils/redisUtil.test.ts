import redisUtil from './redisUtil';

test('onMessage orderBooks', () => {
	const handleOrderBookUpdate = jest.fn();
	redisUtil.onOrderBookUpdate(handleOrderBookUpdate);
	const handleOrderUpdate = jest.fn();
	redisUtil.onOrderUpdate(handleOrderUpdate);
	const handleTradeUpdate = jest.fn();
	redisUtil.onTradeUpdate(handleTradeUpdate);
	redisUtil.onMessage('orderBooks|any', JSON.stringify('test'));
	expect(handleOrderBookUpdate.mock.calls).toMatchSnapshot();
	expect(handleOrderUpdate).not.toBeCalled();
	expect(handleTradeUpdate).not.toBeCalled();
});

test('onMessage orders', () => {
	const handleOrderBookUpdate = jest.fn();
	redisUtil.onOrderBookUpdate(handleOrderBookUpdate);
	const handleOrderUpdate = jest.fn();
	redisUtil.onOrderUpdate(handleOrderUpdate);
	const handleTradeUpdate = jest.fn();
	redisUtil.onTradeUpdate(handleTradeUpdate);
	redisUtil.onMessage('orders|any', JSON.stringify('test'));
	expect(handleOrderUpdate.mock.calls).toMatchSnapshot();
	expect(handleOrderBookUpdate).not.toBeCalled();
	expect(handleTradeUpdate).not.toBeCalled();
});

test('onMessage trades', () => {
	const handleOrderBookUpdate = jest.fn();
	redisUtil.onOrderBookUpdate(handleOrderBookUpdate);
	const handleOrderUpdate = jest.fn();
	redisUtil.onOrderUpdate(handleOrderUpdate);
	const handleTradeUpdate = jest.fn();
	redisUtil.onTradeUpdate(handleTradeUpdate);
	redisUtil.onMessage('trades|any', JSON.stringify('test'));
	expect(handleOrderUpdate).not.toBeCalled();
	expect(handleOrderBookUpdate).not.toBeCalled();
	expect(handleTradeUpdate.mock.calls).toMatchSnapshot();
});

test('onMessage anything else', () => {
	const handleOrderBookUpdate = jest.fn();
	redisUtil.onOrderBookUpdate(handleOrderBookUpdate);
	const handleOrderUpdate = jest.fn();
	redisUtil.onOrderUpdate(handleOrderUpdate);
	redisUtil.onMessage('any', JSON.stringify('test'));
	expect(handleOrderUpdate).not.toBeCalled();
	expect(handleOrderBookUpdate).not.toBeCalled();
});

test('publish', async () => {
	redisUtil.redisPub = null;
	expect(await redisUtil.publish('channel', 'msg')).toBe(0);
	redisUtil.redisPub = {
		publish: jest.fn()
	} as any;
	await redisUtil.publish('channel', 'msg');
	expect(((redisUtil.redisPub as any).publish as jest.Mock).mock.calls).toMatchSnapshot();
});

test('increment', async () => {
	redisUtil.redisPub = null;
	expect(await redisUtil.increment('key')).toBe(0);
	redisUtil.redisPub = {
		incr: jest.fn()
	} as any;
	await redisUtil.increment('key');
	expect(((redisUtil.redisPub as any).incr as jest.Mock).mock.calls).toMatchSnapshot();
});

test('set', async () => {
	redisUtil.redisPub = null;
	expect(await redisUtil.set('key', 'value')).toBe('');
	redisUtil.redisPub = {
		set: jest.fn()
	} as any;
	await redisUtil.set('key', 'value');
	expect(((redisUtil.redisPub as any).set as jest.Mock).mock.calls).toMatchSnapshot();
});

test('get', async () => {
	redisUtil.redisPub = null;
	expect(await redisUtil.get('key')).toBe('');
	redisUtil.redisPub = {
		get: jest.fn()
	} as any;
	await redisUtil.get('key');
	expect(((redisUtil.redisPub as any).get as jest.Mock).mock.calls).toMatchSnapshot();
});

test('hashSet', async () => {
	redisUtil.redisPub = null;
	expect(await redisUtil.hashSet('key', 'field', 'value')).toBe(0);
	redisUtil.redisPub = {
		hset: jest.fn()
	} as any;
	await redisUtil.hashSet('key', 'field', 'value');
	expect(((redisUtil.redisPub as any).hset as jest.Mock).mock.calls).toMatchSnapshot();
});

test('hashGet', async () => {
	redisUtil.redisPub = null;
	expect(await redisUtil.hashGet('key', 'field')).toBe('');
	redisUtil.redisPub = {
		hget: jest.fn()
	} as any;
	await redisUtil.hashGet('key', 'field');
	expect(((redisUtil.redisPub as any).hget as jest.Mock).mock.calls).toMatchSnapshot();
});

test('hashMultiGet', async () => {
	redisUtil.redisPub = null;
	expect(await redisUtil.hashMultiGet('key', 'field')).toEqual({});
	redisUtil.redisPub = {
		hmget: jest.fn(() => Promise.resolve([]))
	} as any;
	await redisUtil.hashMultiGet('key', 'field');
	expect(((redisUtil.redisPub as any).hmget as jest.Mock).mock.calls).toMatchSnapshot();
});

test('hashGetAll', async () => {
	redisUtil.redisPub = null;
	expect(await redisUtil.hashGetAll('key')).toEqual({});
	redisUtil.redisPub = {
		hgetall: jest.fn()
	} as any;
	await redisUtil.hashGetAll('key');
	expect(((redisUtil.redisPub as any).hgetall as jest.Mock).mock.calls).toMatchSnapshot();
});

test('hashDelete', async () => {
	redisUtil.redisPub = {
		hdel: jest.fn()
	} as any;
	await redisUtil.hashDelete('key', 'field');
	expect(((redisUtil.redisPub as any).hdel as jest.Mock).mock.calls).toMatchSnapshot();
});

test('subscribe', async () => {
	redisUtil.redisSub = {
		subscribe: jest.fn()
	} as any;
	redisUtil.subscribe('channel');
	expect(((redisUtil.redisSub as any).subscribe as jest.Mock).mock.calls).toMatchSnapshot();
});

test('patternSubscribe', async () => {
	redisUtil.redisSub = {
		psubscribe: jest.fn()
	} as any;
	redisUtil.patternSubscribe('channel');
	expect(((redisUtil.redisSub as any).psubscribe as jest.Mock).mock.calls).toMatchSnapshot();
});

test('unsubscribe', async () => {
	redisUtil.redisSub = {
		unsubscribe: jest.fn()
	} as any;
	redisUtil.unsubscribe('channel');
	expect(((redisUtil.redisSub as any).unsubscribe as jest.Mock).mock.calls).toMatchSnapshot();
});

test('patternUnsubscribe', async () => {
	redisUtil.redisSub = {
		punsubscribe: jest.fn()
	} as any;
	redisUtil.patternUnsubscribe('channel');
	expect(((redisUtil.redisSub as any).punsubscribe as jest.Mock).mock.calls).toMatchSnapshot();
});

test('push', async () => {
	redisUtil.redisPub = {
		lpush: jest.fn()
	} as any;
	await redisUtil.push('key', 'field');
	expect(((redisUtil.redisPub as any).lpush as jest.Mock).mock.calls).toMatchSnapshot();
});

test('putBack', async () => {
	redisUtil.redisPub = {
		rpush: jest.fn()
	} as any;
	await redisUtil.putBack('key', 'field');
	expect(((redisUtil.redisPub as any).rpush as jest.Mock).mock.calls).toMatchSnapshot();
});

test('pop', async () => {
	redisUtil.redisPub = null;
	expect(await redisUtil.pop('key')).toBe('');
	redisUtil.redisPub = {
		rpop: jest.fn()
	} as any;
	await redisUtil.pop('key');
	expect(((redisUtil.redisPub as any).rpop as jest.Mock).mock.calls).toMatchSnapshot();
});

test('getQueueLength', async () => {
	redisUtil.redisPub = null;
	expect(await redisUtil.getQueueLength('key')).toBe(0);
	redisUtil.redisPub = {
		llen: jest.fn()
	} as any;
	await redisUtil.getQueueLength('key');
	expect(((redisUtil.redisPub as any).llen as jest.Mock).mock.calls).toMatchSnapshot();
});
