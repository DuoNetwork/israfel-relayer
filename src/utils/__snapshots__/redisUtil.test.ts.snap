// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get 1`] = `
Array [
  Array [
    "key",
  ],
]
`;

exports[`getQueueLength 1`] = `
Array [
  Array [
    "key",
  ],
]
`;

exports[`hashDelete 1`] = `
Array [
  Array [
    "key",
    "field",
  ],
]
`;

exports[`hashGet 1`] = `
Array [
  Array [
    "key",
    "field",
  ],
]
`;

exports[`hashGetAll 1`] = `
Array [
  Array [
    "key",
  ],
]
`;

exports[`hashMultiGet 1`] = `
Array [
  Array [
    "key",
    "field",
  ],
]
`;

exports[`hashSet 1`] = `
Array [
  Array [
    "key",
    "field",
    "value",
  ],
]
`;

exports[`increment 1`] = `
Array [
  Array [
    "key",
  ],
]
`;

exports[`onMessage orderBooks 1`] = `
Array [
  Array [
    "orderBooks|any",
    "test",
  ],
]
`;

exports[`onMessage orders 1`] = `
Array [
  Array [
    "orders|any",
    "test",
  ],
]
`;

exports[`onMessage trades 1`] = `
Array [
  Array [
    "trades|any",
    "test",
  ],
]
`;

exports[`patternSubscribe 1`] = `
Array [
  Array [
    "channel",
  ],
]
`;

exports[`patternUnsubscribe 1`] = `
Array [
  Array [
    "channel",
  ],
]
`;

exports[`pop 1`] = `
Array [
  Array [
    "key",
  ],
]
`;

exports[`publish 1`] = `
Array [
  Array [
    "channel",
    "msg",
  ],
]
`;

exports[`push 1`] = `
Array [
  Array [
    "key",
    Array [
      "field",
    ],
  ],
]
`;

exports[`putBack 1`] = `
Array [
  Array [
    "key",
    Array [
      "field",
    ],
  ],
]
`;

exports[`set 1`] = `
Array [
  Array [
    "key",
    "value",
  ],
]
`;

exports[`subscribe 1`] = `
Array [
  Array [
    "channel",
  ],
]
`;

exports[`unsubscribe 1`] = `
Array [
  Array [
    "channel",
  ],
]
`;
